src/Constraint/Constraint.php

```diff
@@ -128,13 +151,18 @@ public function versionCompare($a, $b, $operator, $compareBranches = false)
     }
 
     /**
-     * @param Constraint $provider
+     * @param ConstraintInterface $provider
      * @param bool $compareBranches
      *
      * @return bool
      */
-    public function matchSpecific(Constraint $provider, $compareBranches = false)
+    public function matches(ConstraintInterface $provider, $compareBranches = false)
     {
+        if ($provider instanceof MultiConstraint) {
+            // turn matching around to find a match
+            return $provider->matches($this);
+        }
```

Comment:
The only use case I've found for EmptyConstraint was in Pool;

```
src/Composer/DependencyResolver/Pool.php:330:            $requireFilter = new EmptyConstraint;
```

Not sure what it is used for. Will dig a bit deeper.